students/etwum/lesson04/mailroom_part2.py

```diff
@@ -0,0 +1,153 @@
+                x["avg donation"] = '{:.2f}'.format(avg_donation)
+
+            email = "Dear {a},\n\nThank you for your generous donations of ${b:.2f} to our charity.\n".format(
+                a=x["donor name"],b=x["total donations"])
+            print(email)
+            break
+
+    # if the donor's name is not in the list it adds the name to the list and ask for a donation amount
+    # then prints an email to the new donor
+    else:
+
+        print("Donor not found")
+        new_donation_amount = None
+        boolValid = False
+        while boolValid != True:
+            try:
+                new_donation_amount = float(input("Please input a donation "
+                                                  "amount in order to add the donor to the list. "))
+            except ValueError as e:
+                print(e)
+                print("Please input a valid number")
+            else:
+                boolValid = True
+        new_donor.append({"donor name": donor_full_name, "total donations": new_donation_amount, "number donations": 1,
+                          "avg donation": new_donation_amount})
+        lst_Donors.extend(new_donor)
+        print('-----------------------------------------------------')
+        for x in lst_Donors:
+            if x["donor name"] == donor_full_name:
+                email = "Dear {a},\n\nThank you for your generous donations of ${b:.2f} to our charity.\n".format(
+                    a=x["donor name"],b=x["total donations"])
+                print(email)
+    print('-----------------------------------------------------')
+
+
+def sort_list(lst_Donors):
+    # used to sort the donor list by the total donations column
+
+    return lst_Donors["total donations"]
+
+
+def create_report():
+    # creates a report of the the donors
+    # headers used in table
+    print()
+    lst_Header = [["Donor Name", "| Total Donation(s)", "| # of Donations", "| Avg Donation"]]
+
+    for x in lst_Header:
+        print('{:<25}{:<20}{:<17}{:<15}'.format(*x))
+    print("----------------------------------------------------------------------------")
+    for x in sorted(lst_Donors,key=sort_list, reverse= True):
+        print('{:<25} $ {:<20}{:^14} $ {:<15}'.format(*x.values()))
+
+
+def send_letter_all():
+    print('\n----------------------------------------------')
+    file = ()
+    # creates a letter to every donor by writing them to separate text files
+    for x in lst_Donors:
+        # opens or creates a new text file for writing based on the donor name
+        file = open('{a}.txt'.format(a = x['donor name']), 'w')
+
+        # writes a thank you letter to the text file
+        file.write('Dear {a},\n\nThank you for your generous donation(s) of ${b:.2f} to our charity.'.format(
+            a=x['donor name'], b=x['total donations']))
+        print('Letter sent to  {a}'.format(a=x['donor name']))
+
+    #closes and saves the file
+    file.close()
```

Comment:
you are only closing the last file you work on, because you creating multiple `file` instances for each donor.

you should use `with` statement, which takes care of closing the file for you:
```
with open('filename', 'w') as fh:
    fh.write('stuff')
```

Reply:
try using more descriptive variables rather than `a` and `b`